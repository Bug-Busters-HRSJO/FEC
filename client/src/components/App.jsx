import React from 'react';
import Atelier from '../Atelier.js';
import Overview from './Overview.jsx';
import Ratings from './RatingsReview.jsx';
import RICWidget from './RIC-Widget.jsx';
import Header from './Header.jsx';
import Footer from './Footer.jsx';


class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      product: '13060',
      stylePath: 'lightTheme.css',
      cart: []
    }

    this.handleProductHighlight = this.handleProductHighlight.bind(this);
    this.toggleTheme = this.toggleTheme.bind(this);
    this.handleProductClick = this.handleProductClick.bind(this);

  }

  handleProductClick(id) {
    this.setState({ product: id });
  }

  handleProductHighlight() {
    this.setState({ product: '13357'});
  }

  loadUserCart () {
    const local = window.localStorage;
    // load user cart if one exists already
  }

  toggleTheme() {
    this.setState(state => {
      return {
        stylePath:
          state.stylePath === 'darkTheme.css'
          ? 'lightTheme.css'
          : 'darkTheme.css'
      }
    });

    const local = window.localStorage;

    this.state.stylePath === 'darkTheme.css'
    ? local.setItem('theme', 'lightTheme.css')
    : local.setItem('theme', 'darkTheme.css');
  }

  componentDidMount() {
    const local = window.localStorage;

    local.getItem('theme')
    ? this.setState({stylePath: local.getItem('theme')})
    : local.setItem('theme', this.state.stylePath);
  }

  render() {
    const {
      product,
      stylePath
    } = this.state;

    return (
      <>
        <link rel="stylesheet" type="text/css" href={stylePath} />
        <Header
          highlight={this.handleProductHighlight}
          theme={stylePath}
          toggleTheme={this.toggleTheme}/>
        {/* <Overview
        theme={ stylePath }
        productId={ this.state.product }
        /> */}
        <Ratings id={this.state.product} />
        {/* <RICWidget
          productId={product}
          productClick={this.handleProductClick}/> */}
        <Footer />
      </>
    );
  }
};


export default App;