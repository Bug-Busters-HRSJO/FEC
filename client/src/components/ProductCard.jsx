import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { auth } from '../../../config.js';

// take in destructing {productId}
function ProductCard() {
  const productId = 13027; //placeholder for testing
  // const []

  async function fetchProducts() {
    let productData = await axios.get(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-sjo/products/${productId}`,
    { headers: { 'Authorization': auth.TOKEN } });

    let productStyles = await axios.get(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-sjo/products/${productId}/styles`,
    { headers: { 'Authorization': auth.TOKEN } });

    console.log(productData.data);
    console.log(productStyles.data);
  };

  useEffect(() => {
    fetchProducts().catch((err) => console.log(`Error fetching product info: ${err}`))
  }, []);

  return (
    <div>
      <h1>Testing-ProductCard</h1>
      <div></div>
    </div>
  );
};

export default ProductCard;

/*

- pull productId from relatedList

1) product preview image
2) product category
3) product price
4) star rating

*/