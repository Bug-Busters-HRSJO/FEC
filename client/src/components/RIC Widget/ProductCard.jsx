import { useState, useEffect } from 'react';
import axios from 'axios';
import { auth } from '../../../../config.js';
import AvgRating from '../AvgRating.jsx';

function ProductCard({ productId }) {
  const [image, setImage] = useState('');
  const [category, setCategory] = useState('');
  const [name, setName] = useState('');
  const [price, setPrice] = useState(0);
  const [salePrice, setSale] = useState(0);

  async function fetchProducts() {
    let productData = await axios.get(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-sjo/products/${productId}`,
    { headers: { 'Authorization': auth.TOKEN } });

    let productStyles = await axios.get(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-sjo/products/${productId}/styles`,
    { headers: { 'Authorization': auth.TOKEN } });

    const product = productData.data;
    const firstStyle = productStyles.data.results[0];

    setCategory(product.category);
    setName(product.name);
    setImage(firstStyle.photos[0].thumbnail_url);
    setPrice(firstStyle.original_price);

    if (firstStyle.sale_price !== null) {
      setSale(firstStyle.sale_price);
    };
  };

  useEffect(() => {
    fetchProducts().catch((err) => console.log(`Error fetching product info: ${err}`))
  }, [productId]);

  return (
    <div>
      <div>
        <img src={image} alt={name}></img>
        <div>{category}</div>
        <div>{name}</div>
        <div>${price}</div>
        <AvgRating
          productId={productId}/>
      </div>
    </div>
  );
};

export default ProductCard;

/*

- pull productId from relatedList

1) product preview image
2) product category
3) product name
4) product price
5) star rating

*/