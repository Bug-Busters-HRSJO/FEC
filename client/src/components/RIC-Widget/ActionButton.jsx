import React, { useState, useEffect }  from 'react';

const ActionButton = ({ id, index, listState, triggerDelete, triggerModal, offset }) => {
  const [isStar, setStar] = useState(true);

  const setAction = () => {
    listState === 'related'
    ? setStar(true)
    : setStar(false);
  };

  useEffect(() => {
    setAction();
  }, [listState]);

  return (
    <span className='ActionButton'>
      {isStar
        ? <i data-testid='star' className="fas fa-star" id='ActButton' key={id} onClick={() => {triggerModal(id)}}></i>
        : <i data-testid='x-button' className="fas fa-times" id='ActButton' key={index} type="button" value='âœ–' onClick={() => {triggerDelete(index)}}></i>
      }
    </span>
  )
};

export default ActionButton;