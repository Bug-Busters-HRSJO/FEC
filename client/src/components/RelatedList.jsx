/*
- get product Id from overview design product to then map over that list


List Behavior

- list is determined internally & related to current product view
- # of related products finite
- displayed in a carousel fashion
 - scrolling horizontally
 - first item on left hand side
 - arrows appear on left/right
  - click to scroll thru list
  - only one item per scroll click
  - left arrow is hidden if at first item at far lest
  - right arrow hidden if last item at far right
 - any prod not on screen should appear offscreen in carousel
*/