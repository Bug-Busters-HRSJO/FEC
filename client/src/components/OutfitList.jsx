import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { auth } from '../../../config.js';
import ProductCard from './ProductCard.jsx';

// expecting this productId to be linked to current Overview Product
function OutfitList({ productId }) {
  const [ifOutfit, setExists] = useState(false);
  const [outfitItems, setOutfit] = useState([]);

  function addItem() {
    if (productId !== 0 && outfitItems.indexOf(productId) === -1) {
      setOutfit(prevArray => [...prevArray, productId]);
      setExists(true);
    }
  };

  useEffect(() => {
    addItem();
  }, [productId]);

  return (
    <div>
      <h1>Testing-OutfitList</h1>
      <div className='AddCard'>
        <div>+</div>
        <div>Add to Outfit</div>
      </div>
      <div>
        {ifOutfit
          ? <div> {outfitItems.map((id, index) => (
            <ProductCard
              productId={id} key={index}/>))} </div>
          : null
        }
      </div>
    </div>
  );

};

export default OutfitList;

/*
When clicking 'Add to Outfit', pull the productId of Overview and add productcard
- on click, add the productid to an outfit array
- render the product card of that product id




- custom product list created by user w/ products that they've grouped together for 'outfit'
- similar format to related product cards
- will not be determined internally but unique to user
 - items only added when explicitly added by user
- default, no items
- first card should not contain a product, should have '+' icon and read 'add to outfit'
 - this remains visible even as items are added
- only unique to specific user, same across the board, regardless where they click to
- product can only be added once
- no max # of items to this list
- PERSIST
 - across page navigation
 - even after user has exited site and return later
*/