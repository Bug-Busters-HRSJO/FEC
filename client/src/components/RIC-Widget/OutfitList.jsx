import React, { useState } from 'react';
import ProductCard from './ProductCard.jsx';

function OutfitList({ productId }) {
  const [ifOutfit, setExists] = useState(false);
  const [outfitItems, setOutfit] = useState([]);

  function addItem(id) {
    if (outfitItems.indexOf(id) === -1) {
      setOutfit(prevArray => [...prevArray, id]);
      setExists(true);
    }
  };

  return (
    <div>
      <div className='AddCard' onClick={() => addItem(productId)}>
        <div>+</div>
        <div>Add to Outfit</div>
      </div>
      <div>
        {ifOutfit
          ? <div> {outfitItems.map((id, index) => (
            <ProductCard
              productId={id} key={index}/>))} </div>
          : null
        }
      </div>
    </div>
  );

};

export default OutfitList;

/*
When clicking 'Add to Outfit', pull the productId of Overview and add productcard
- on click, add the productid to an outfit array
- render the product card of that product id




- custom product list created by user w/ products that they've grouped together for 'outfit'
- similar format to related product cards
- will not be determined internally but unique to user
 - items only added when explicitly added by user
- default, no items
- first card should not contain a product, should have '+' icon and read 'add to outfit'
 - this remains visible even as items are added
- only unique to specific user, same across the board, regardless where they click to
- product can only be added once
- no max # of items to this list
- PERSIST
 - across page navigation
 - even after user has exited site and return later
*/