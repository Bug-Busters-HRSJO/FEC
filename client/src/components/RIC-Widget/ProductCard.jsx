import React, { useState, useEffect } from 'react';
import AvgRating from '../AvgRating.jsx';
import ActionButton from './ActionButton.jsx';
import Atelier from '../../Atelier.js';
import errimage from '../../imgs/imagenot.png';
import Gallery from './PCGallery.jsx';

const ProductCard = ({ productId, index, listState, triggerDelete, triggerModal, offset, productClick }) => {
  const [image, setImage] = useState('');
  const [category, setCategory] = useState('');
  const [name, setName] = useState('');
  const [price, setPrice] = useState({ default: 0, salePrice: null });
  const [gallery, setGalley] = useState([]);
  const [showGallery, setShow] = useState(false);

  const fetchProducts = async () => {
    let productData = await Atelier.getInfo(productId);
    let productStyles = await Atelier.getStyles(productId);

    const firstStyle = productStyles.results[0];

    setCategory(productData.category);
    setName(productData.name);
    setImage(firstStyle.photos[0].thumbnail_url);
    setPrice({
      default: firstStyle.original_price,
      salePrice: firstStyle.sale_price
      });

    setGalley(firstStyle.photos);
  };

  const mouseOver = () => {
    setShow(true);
  }

  const mouseOut = () => {
    setShow(false);
  }

  const changeImage = (index) => {
    setImage(gallery[index].thumbnail_url);
  }

  useEffect(() => {
    fetchProducts().catch((err) => console.log(`Error fetching product/style info: ${err}`));
  }, [productId]);

  return (
    <span className='ProductCard' style={{ '--offset': offset }}>
      <ActionButton
        index={index}
        id={productId}
        listState={listState}
        triggerDelete={triggerDelete}
        triggerModal={triggerModal}/>

        <div className='product-gallery'>
          <img className='ProductImage'
               src={image ? image : errimage}
               alt={name}
               onMouseOver={mouseOver}
               onClick={() => productClick(productId)}
               data-testid='image-prod'>
          </img>
          {showGallery
            ? <>
                <div className='image-holder'
                     onMouseOut={mouseOut}
                     onClick={() => productClick(productId)}
                     data-testid='image-holder'></div>
                <Gallery
                  gallery={gallery}
                  changeImage={changeImage}/>
              </>
            : <span data-testid='no-gallery' hidden></span>
          }
        </div>
        <div data-testid='productcard' onClick={() => productClick(productId)}>
        <span id='prod-category'>{category}</span>
        <br />
        <b id='prod-name'>{name}</b>
        <br />
        {price.salePrice
          ? <> <b id="sale-price">${price.salePrice} &nbsp; </b> <strike className='prod-price'>${price.default}</strike> </>
          : <span className='prod-price'>${price.default}</span>
        }
        <br />
        <span id='prod-star'>
          <AvgRating
            productId={productId}/>
        </span>
      </div>
    </span>
  );
};

export default ProductCard;

/*
	Upon hovering over the initial preview image that appears on the card, a carousel of thumbnail images should appear allowing the user to view additional preview images.
  The carousel will be a row of thumbnails which overlays the bottom portion of the preview image, covering part of the image when hovered.
  The list should disappear when the user is no longer hovering on the card.
	Four thumbnail images should appear in the carousel at any given time.
  The carousel should extend the ability to scroll left/right through the list via left and right arrows such that all images for that product can be selected and viewed

*/