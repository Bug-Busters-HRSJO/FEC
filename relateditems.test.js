import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';

import App from './client/src/components/App.jsx';
import ActionButton from './client/src/components/RIC-Widget/ActionButton.jsx';
import OutfitList from './client/src/components/RIC-Widget/OutfitList.jsx';
import ProductCard from './client/src/components/RIC-Widget/ProductCard.jsx';
import RelatedList from './client/src/components/RIC-Widget/RelatedList.jsx';

// basic render testing, screen.debug() will log the HTML output in test terminal
describe('App', () => {
  test('renders App component', () => {
    render(<App />);
  });
});

describe('Action Button', () => {
  test('star action button exists on initial render', () => {
    render(<ActionButton />);

    expect(screen.getByDisplayValue('â˜†')).toBeInTheDocument();
  });
});

describe('Action Button', () => {
  test('delete action button does not exist on initial render', () => {
    render(<ActionButton />);

    expect(screen.queryByDisplayValue('x')).toBeNull();
  });
});

describe('Action Button', () => {
  test('calls the onClick callback handler', async () => {
    const onClick = jest.fn();

    render( <input type="button" value='x' onClick={onClick} /> );

    await userEvent.click(screen.getByDisplayValue('x'));

    expect(onClick).toHaveBeenCalledTimes(1);
  });
});

describe('Add to Outfit card showing up', () => {
  test('check for outfit text', () => {
    render(<OutfitList />);

    expect(screen.getByText(/Add to Outfit/)).toBeInTheDocument();
  });
});

describe('Outfit List', () => {
  test('calls the onClick callback handler', async () => {
    const onClick = jest.fn();

    render( <div className='AddCard' onClick={onClick}>
      <div data-testid='outfit'>Add to Outfit</div>
    </div> );

    await userEvent.click(screen.getByTestId('outfit'));

    expect(onClick).toHaveBeenCalledTimes(1);
  });
});

// checking rendering for now, need to work on better test for this
describe('Product Card', () => {
  test('check existance', () => {
    render(<ProductCard />);
  });
});

// checking rendering for now, need to work on better test for this
describe('Related List', () => {
  test('check existance', () => {
    render(<RelatedList />);
  });
});