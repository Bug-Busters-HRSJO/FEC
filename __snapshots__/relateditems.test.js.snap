// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RICWidget renders RIC-Widget component 1`] = `
<DocumentFragment>
  <div
    id="RICWid"
  >
    <div>
      <h3>
        Related Products
      </h3>
      <div
        class="RICList"
      />
    </div>
    <div>
      <h3>
        Your Outfit
      </h3>
      <div
        class="RICList"
      >
        <div
          class="AddCard"
        >
          <div>
            <div>
              âŠ•
            </div>
            <div>
              Add to Outfit
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
